pkg/gardenlet/controller/shoot/shoot_control_prepare_migration.go

```diff
@@ -0,0 +1,298 @@
+		func(errorID string) error {
+			o.CleanShootTaskError(context.TODO(), errorID)
+			return nil
+		},
+		nil,
+		utilerrors.ToExecute("Create botanist", func() error {
+			return retryutils.UntilTimeout(context.TODO(), 10*time.Second, 10*time.Minute, func(context.Context) (done bool, err error) {
+				botanist, err = botanistpkg.New(o)
+				if err != nil {
+					return retryutils.MinorError(err)
+				}
+				return retryutils.Ok()
+			})
+		}),
+	)
+
+	if err != nil {
+		return gardencorev1beta1helper.NewWrappedLastErrors(gardencorev1beta1helper.FormatLastErrDescription(err), err)
+	}
+
+	var (
+		defaultTimeout  = 10 * time.Minute
+		defaultInterval = 5 * time.Second
+
+		g = flow.NewGraph("Shoot's control plane preparation for migration")
+
+		deploySecrets = g.Add(flow.Task{
+			Name: "Deploying Shoot certificates / keys",
+			Fn:   flow.TaskFn(botanist.DeploySecrets),
+		})
+		deployBackupEntryInGarden = g.Add(flow.Task{
+			Name: "Deploying backup entry",
+			Fn:   flow.TaskFn(botanist.DeployBackupEntryInGarden),
+		})
+		wailtUntilBackupEntryInGardenReconciled = g.Add(flow.Task{
+			Name:         "Waiting until the backup entry has been reconciled",
+			Fn:           flow.TaskFn(botanist.WaitUntilBackupEntryInGardenReconciled),
+			Dependencies: flow.NewTaskIDs(deployBackupEntryInGarden),
+		})
+		deployETCD = g.Add(flow.Task{
+			Name:         "Deploying main and events etcd",
+			Fn:           flow.TaskFn(botanist.DeployETCD).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(deploySecrets, wailtUntilBackupEntryInGardenReconciled),
+		})
+		waitUntilEtcdReady = g.Add(flow.Task{
+			Name:         "Waiting until main and event etcd report readiness",
+			Fn:           flow.TaskFn(botanist.WaitUntilEtcdReady),
+			Dependencies: flow.NewTaskIDs(deployETCD),
+		})
+		prepareControlPlaneDeployments = g.Add(flow.Task{
+			Name:         "Stop Kubernetes API Server in Shoot's namespace",
+			Fn:           flow.TaskFn(botanist.PrepareControlPlaneDeployments).RetryUntilTimeout(defaultInterval, defaultTimeout).SkipIf(shootHibernated),
+			Dependencies: flow.NewTaskIDs(waitUntilEtcdReady),
+		})
+		waitUntilAPIServerIsDown = g.Add(flow.Task{
+			Name:         "Wait for Kubernetes API Server to be scaled to zero",
+			Fn:           flow.TaskFn(botanist.WaitUntilKubeAPIServerScalesDown).SkipIf(shootHibernated),
+			Dependencies: flow.NewTaskIDs(prepareControlPlaneDeployments),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying ingress DNS record",
+			Fn:           flow.TaskFn(botanist.DestroyIngressDNSRecord).DoIf(dnsEnabled),
+			Dependencies: flow.NewTaskIDs(waitUntilAPIServerIsDown),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying external domain DNS record",
+			Fn:           flow.TaskFn(botanist.DestroyExternalDomainDNSRecord).DoIf(dnsEnabled),
+			Dependencies: flow.NewTaskIDs(waitUntilAPIServerIsDown),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying internal domain DNS record",
```

Comment:
What is the the proper migration plan for the DNS records? I guess we can't keep it like this, at least for the ingress record, as this might result in a downtime of applications running in the shoot. cc @mandelsoft

Reply:
@swilen-iwanow I discussed yesterday with @mandelsoft and @MartinWeindel and the way to go is the `DNSOwner` object in the `dns.gardener.cloud/v1alpha1` API group. For this PR we can keep it like this for now to not unnecessarily block it as the API validation still forbids that `.spec.seedName` gets changed, so this code cannot be triggered by end-users. However, could you please get in touch with the two colleagues and file a separate PR that introduces the DNS owner objects (maybe first outline a plan, open an issue and describe it, and then work on the PR)?

Reply:
OK will do after I finish my second change.